### PR TITLE
Leads managment

### PR DESCRIPTION
**Pull request by snebo**
## Description

Introduces lead management to website API list. This handles creating a new temporary lead application in the database and modifying a use when the application is approved, admin management for lead applications, generating unique links for the lead application.

> A new temporary lead registration collection was created. I felt it might be more optimal to have the applications in a separate collection.
> 

> Included crypt for decoding and encoding the generated link -for security-
> 

> Included date_dtf to render our date properly
> 

> Slightly modified user schema to accommodate for certain logics.
> 

### Issues:

- When handling the generated links in swagger, it can not find the links, because it is searching for a GET, while the links are POST. I don’t know if there’s a way to change this without changing swagger config files

### Type of change
Bug fix (removed unnecessary relinking to database previously added by me in app module)

New feature 

- Admins can Generate invitation links for existing and new users
- Admin can approve or delete, view all or view one temp lead application.
- Existing users can apply for a lead position once every 3 months
- Encryption and decryption of URLs
- Modified utils to include encrypt and decrypt functions
- Added API to view all leads

## Checklist:

- Style guidelines of this project
    - Follow the naming and style guidelines and prettier format
    - Reviewed and removed duplication
    - Commented code
    - Changes generate no new warning
- Leads management
    - [x]  Users can apply to be leads
    - [x]  admins can view one application
    - [x]  admins can view all applications
    - [x]  admins can approve an application
    - [x]  admins can reject an application
    - [x]  admins can generate a link for registration
        - [x]  the link creates a new user and a lead application for new users
        - [x]  the link creates a lead application for existing users
    - [x]  list all verified leads

### CODE REVIEW

**Comment by snebo** [comment]
Handlebars templates would need to be refined made for some

File: src/app.module.ts [hunk 1]
```diff
@@ -8,6 +8,8 @@ import { AuthModule } from './shared/auth/auth.module';
 import { DataLogsModule } from './shared/datalogs';
 import { UsersModule } from './users/users.module';
 import { EventModule } from './events/events.users.module';
+// import { NodeMailService } from './node-mail/node-mail.service';
+// import { NodeMailModule } from './node-mail/node-mail.module';
```
Comment (willypelz) [inline src/app.module.ts:12]: if this comment section is not used, we can get it cleaned up from the codebase

File: src/app.module.ts [hunk 2]
```diff
@@ -21,6 +23,7 @@ import { EventModule } from './events/events.users.module';
     DataLogsModule,
     UsersModule,
     EventModule,
```
Comment (willypelz) [inline src/app.module.ts:25]: NodeMailService needs to be cleaned up from the project

File: src/app.module.ts [hunk 3]
```diff
@@ -29,6 +32,7 @@ import { EventModule } from './events/events.users.module';
     ThrottlerGuard,
     LogInterceptor,
     AllExceptionsFilter,
+    // NodeMailService,
```
Comment (willypelz) [inline src/app.module.ts:35]: same here

File: src/shared/auth/guards/jwt.admins.guard.ts [hunk 4]
```diff
@@ -21,11 +21,13 @@ export class JwtAdminsGuard
   async canActivate(context: ExecutionContext): Promise<boolean> {
     const request = context.switchToHttp().getRequest();
     const token = this.extractTokenFromHeader(request);
+    console.log(token);
     if (!token) {
       throw new UnauthorizedException();
     }
     try {
       const payload = await global.jwtService.verify(token);
+      console.log(`user role: ${payload.role}`);
```
Comment (willypelz) [inline src/shared/auth/guards/jwt.admins.guard.ts:30]: you can remove the debugging line of code

**Comment by snebo** [comment]
SMTP_HOST=smtp.mailtrap.io
SMTP_PORT=2525
SMTP_USER=4b3474c62b2e5e
SMTP_PASS=25271685cb1a3e

would be required in the env file to test the mailing service. a mailtrap free account can be easily created and comes with a user and pass

File: src/shared/utils/index.ts [hunk 5]
```diff
@@ -5,7 +5,8 @@ export * from './cloudinary.util';
 export * from './paging.util';
 export * from './redact.util';
 export * from './access-token-validator.util';
-export * from './mailgun.util';
+// export * from './mailgun.util';
```
Comment (willypelz) [inline src/shared/utils/index.ts:8]: same here

File: src/users/users.admin.controller.ts [hunk 6]
```diff
@@ -10,23 +10,31 @@ import {
...
+} from '@nestjs/swagger';
 import { UsersService } from './users.service';
 import { User, UserDocument } from 'src/shared/schema';
 import { Model } from 'mongoose';
 import { JwtAdminsGuard } from 'src/shared/auth/guards/jwt.admins.guard';
 import { UserInviteDto } from './dto/user-invite.dto';
 import {
   ApiReq,
-  UserStatus,
+  // UserStatus,
```
Comment (willypelz) [inline src/users/users.admin.controller.ts:29]: same here

File: src/users/users.service.ts [hunk 7]
```diff
@@ -251,15 +257,14 @@ export class UsersService {
     sendMail({
       to: user.email,
       from: EmailFromType.HELLO,
-      subject: 'Password Change',
+      subject: 'PASSWORD CHANGE',
```
Comment (willypelz) [inline src/users/users.service.ts:260]: why all caps, as subject.   'Password Change', looks better

**Review by willypelz — DISMISSED** [review]
pending comments to be fixed

**Comment by snebo** [comment]
changes resolved

**Review by willypelz — APPROVED** [review]
LGTM

**Pull request merged**